### PR TITLE
[Dynamic Neighbor]: add patch to support dynamic neighbor configuration.

### PR DESCRIPTION
**Pull request by sihuihan88**
This patch bases on https://github.com/FRRouting/frr/commit/f14e6fdbe2ab77712bffa9a2de361958634c494d, FRRouting/frr@20eb886, FRRouting/frr@57e9ee0, FRRouting/frr@2aab8d2

To make it work in our code base, the major additional changes are:
1. Add openbsd-queue.h in lib/ 
2. Add bgp_af_index, afindex definition in bgpd/bgpd.h
3. Add peer_af_create(), peer_af_find() in bgpd/bgpd.c
4. Add MTYPE_BGP_PEER_AF in memtypes.c and memtypes.h
5. Change bgp_debug_neighbor_events() to BGP_Debug() in bgpd.c and bgp_fsm.c
6. Minor change of peer_delete() in bgpd.c
7. Minor change of peer_and_group_lookup_vty(), peer_remote_as_vty () in bgp_vty.c
8. Minor change of logging info.

### CODE REVIEW

File: bgpd/bgp_fsm.c [hunk 1]
```diff
@@ -656,6 +689,17 @@ bgp_connect_success (struct peer *peer)
...
+   {
+      if (BGP_DEBUG (events, EVENTS))
+      {
+        zlog_debug ("%s (dynamic neighbor) deleted", peer->host);
+      }
+
+      peer_delete (peer);
+      return -1;
+   }
+
```
Comment (pavel-shirshov) [inline bgpd/bgp_fsm.c:702]: The same code 4 times. Probably it's better to introduce a static function here?

File: bgpd/bgp_network.c [hunk 2]
```diff
@@ -220,6 +220,7 @@ bgp_accept (struct thread *thread)
 
   peer1 = peer_lookup (NULL, &su);
 
+
```
Comment (pavel-shirshov) [inline bgpd/bgp_network.c]: Probably it's better to remove it?

**Review by pavel-shirshov — CHANGES_REQUESTED** [review]

**Comment by stcheng** [comment]
could you provide the patch info you reference? Add it to the commit message and also state what you have changed due to the patch conflicts.

File: bgpd/bgp_fsm.c (hunk 1 above)
Comment (sihuihan88) [reply bgpd/bgp_fsm.c:702]: The code is used to provide log info if debug_event is enabled. They are not exactly the same. So adding more detailed log info now for better troubleshooting.

File: bgpd/bgp_network.c (hunk 2 above)
Comment (sihuihan88) [reply bgpd/bgp_network.c]: Good catch. Have updated accordingly.

**Comment by lguohan** [comment]
can you check this commit? https://github.com/FRRouting/frr/commit/20eb8864bbaf19771502b6c1ed3f1e43186f97eb

**Comment by lguohan** [comment]
and this one. https://github.com/FRRouting/frr/commit/57e9ee0a9430e0fa0c7fe0751eaf0129abb7fa9c

**Comment by lguohan** [comment]
and this one https://github.com/FRRouting/frr/commit/2aab8d2b2cd88163d48317224635c90f0dd3a50d

File: bgpd/bgp_fsm.c [hunk 3]
```diff
@@ -439,6 +439,17 @@ bgp_stop (struct peer *peer)
   safi_t safi;
   char orf_name[BUFSIZ];
 
+  if (peer_dynamic_neighbor(peer) &&
+      !(CHECK_FLAG(peer->flags, PEER_FLAG_DELETE)))
+    {
+      if (BGP_DEBUG (events, EVENTS))
+      {
+          zlog_debug("BGP stop: %s (dynamic neighbor) deleted", peer->host);
```
Comment (lguohan) [inline bgpd/bgp_fsm.c:447]: -> zlog_info

check line 467: neighbor down event is logged as info level, this is also neighbor down event, should also logged as info level.

**Comment by lguohan** [comment]
next time, you need to do a fork and https://help.github.com/articles/fork-a-repo/

File: bgpd/bgp_fsm.c [hunk 4]
```diff
@@ -588,6 +599,17 @@ bgp_stop_with_error (struct peer *peer)
...
+  {
+    if (BGP_DEBUG (events, EVENTS))
+    {
+      zlog_debug ("BGP stop with error: %s (dynamic neighbor) deleted", peer->host);
+    }
+
+    peer_delete (peer);
+    return -1;
+  }
+
```
Comment (lguohan) [inline bgpd/bgp_fsm.c:612]: I think this logic can be merged into bgp_stop() as bgp_stop already has similar logic. We should enable code re-use.

File: bgpd/bgp_fsm.c [hunk 5]
```diff
@@ -609,6 +631,17 @@ bgp_stop_with_notify (struct peer *peer, u_char code, u_char sub_code)
...
+  if (peer_dynamic_neighbor(peer))
+    {
+      if (BGP_DEBUG (events, EVENTS))
+      {
+        zlog_debug ("BGP stop with notify: %s (dynamic neighbor) deleted", peer->host);
+      }
+
+      peer_delete (peer);
+      return -1;
+    }
```
Comment (lguohan) [inline bgpd/bgp_fsm.c:643]: should move into bgp_stop()

File: bgpd/bgp_fsm.c [hunk 6]
```diff
@@ -656,6 +689,17 @@ bgp_connect_success (struct peer *peer)
...
+  if (peer_dynamic_neighbor(peer))
+   {
+      if (BGP_DEBUG (events, EVENTS))
+      {
+        zlog_debug ("BGP connection fail: %s (dynamic neighbor) deleted", peer->host);
+      }
+
+      peer_delete (peer);
+      return -1;
+   }
```
Comment (lguohan) [inline bgpd/bgp_fsm.c:701]: move into bgp_stop().

File: bgpd/bgpd.c [hunk 7]
```diff
@@ -2268,6 +2394,206 @@ peer_lookup_with_open (union sockunion *su, as_t remote_as,
...
+  peer = peer_create_bind_dynamic_neighbor (gbgp, su, group);
+  assert (peer);
+
+  gbgp->dynamic_neighbors_count = ++dncount;
+
+  if (BGP_DEBUG (events, EVENTS))
+  {
+    zlog_debug ("%s Dynamic Neighbor added, group %s count %d",
+                peer->host, group->name, dncount);
+  }
```
Comment (lguohan) [inline bgpd/bgpd.c:2575]: this is neighbor add event, I think it should be info level, we should print it out when we enabled log neighbor change.

File: bgpd/bgpd.c [hunk 8]
```diff
@@ -2268,6 +2394,206 @@ peer_lookup_with_open (union sockunion *su, as_t remote_as,
...
+      dncount = peer->group->bgp->dynamic_neighbors_count;
+      if (dncount)
+        peer->group->bgp->dynamic_neighbors_count = --dncount;
+    }
+
+  if (BGP_DEBUG (events, EVENTS))
+  {
+    zlog_debug ("%s dropped from group %s, count %d",
+                 peer->host, peer->group->name, dncount);
+  }
```
Comment (lguohan) [inline bgpd/bgpd.c:2594]: we should do something simliar here.

      /* bgp log-neighbor-changes of neighbor Down */
      if (bgp_flag_check (peer->bgp, BGP_FLAG_LOG_NEIGHBOR_CHANGES))
 zlog_info ("%%ADJCHANGE: neighbor %s Down %s", peer->host,
                   peer_down_str [(int) peer->last_reset]);

**Review by lguohan — CHANGES_REQUESTED** [review]
mostly logging information related

**Review by lguohan — APPROVED** [review]

**Review by pavel-shirshov — APPROVED** [review]

**Pull request merged**